lib/dor/text_extraction/abbyy/split_alto.rb

```diff
@@ -27,30 +23,46 @@ def output_path(filename)
           File.join(File.dirname(alto_path), filename)
         end
 
+        # NOTE: we avoid using Nokogiri XML Builder to build the XML structure to reduce memory usage
         def xml_structure(page)
-          Nokogiri::XML::Builder.new do |xml|
-            xml.alto(namespaces) do
-              xml.parent << ocr_processing_nodes
-              xml.parent << style_nodes
-              xml.parent << page
-            end
+          "<?xml version=\"1.0\"?>\n<alto #{namespaces}>\n#{ocr_processing_nodes}\n#{style_nodes}\n#{page}\n</alto>\n"
+        end
+
+        # NOTE: do not use Nokogiri to load and parse the XML in memory at once as it is memory intensive
+        # first pass through the XML to get the common nodes
+        def fetch_common_nodes
+          Nokogiri::XML::Reader(File.open(alto_path)).each do |node|
+            @style_nodes = node.outer_xml if node?(node, 'Styles')
+            @ocr_processing_nodes = node.outer_xml if node?(node, 'OCRProcessing')
+            @page_filename_nodes = node.outer_xml if node?(node, 'sourceImageInformation')
```

Comment:
I didn't know about `outer_xml()`, that's a good one to remember.

Reply:
yeah, useful here